kratos/solving_strategies/convergencecriterias/mixed_generic_criteria.h

```diff
@@ -0,0 +1,441 @@
+
+    typedef typename BaseType::TDataType TDataType;
+
+    typedef typename BaseType::DofsArrayType DofsArrayType;
+
+    typedef typename BaseType::TSystemMatrixType TSystemMatrixType;
+
+    typedef typename BaseType::TSystemVectorType TSystemVectorType;
+
+    typedef typename DofUpdater<TSparseSpace >::UniquePointer DofUpdaterPointerType;
+
+    typedef std::vector<std::tuple<VariableData*, TDataType, TDataType>> ConvergenceVariableListType;
+
+    typedef std::size_t KeyType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Constructor.
+
+    /**
+     * @brief Construct a new Mixed Generic Criteria object
+     * Construct the mixed generic convergence criteria from a convergence variables list.
+     * The convergence variable list contains for each variable the variable itself as well as the corresponding relative and absolute tolerances.
+     * @param rConvergenceVariablesList List containing tuples with the convergence variables to be checked. The tuples are set as <Variable, relative tolerance, absolute tolerance>
+     */
+    MixedGenericCriteria(const ConvergenceVariableListType& rConvergenceVariablesList)
+        : ConvergenceCriteria<TSparseSpace, TDenseSpace>()
+        , mVariableSize([&] (const ConvergenceVariableListType& rList) -> int {return rList.size();} (rConvergenceVariablesList))
+        , mVariableDataVector([&] (const ConvergenceVariableListType& rList) -> std::vector<VariableData*> {
+            int i = 0;
+            std::vector<VariableData*> aux_vect(mVariableSize);
+            for (const auto &r_tup : rList) {
+                aux_vect[i++] = std::get<0>(r_tup);
+            }
+            return aux_vect;
+        } (rConvergenceVariablesList))
+        , mRatioToleranceVector([&] (const ConvergenceVariableListType& rList) -> std::vector<TDataType> {
+            int i = 0;
+            std::vector<TDataType> aux_vect(mVariableSize);
+            for (const auto &r_tup : rList) {
+                aux_vect[i++] = std::get<1>(r_tup);
+            }
+            return aux_vect;
+        } (rConvergenceVariablesList))
+        , mAbsToleranceVector([&] (const ConvergenceVariableListType& rList) -> std::vector<TDataType> {
+            int i = 0;
+            std::vector<TDataType> aux_vect(mVariableSize);
+            for (const auto &r_tup : rList) {
+                aux_vect[i++] = std::get<2>(r_tup);
+            }
+            return aux_vect;
+        } (rConvergenceVariablesList))
+        , mLocalKeyMap([&] (const ConvergenceVariableListType& rList) -> std::unordered_map<KeyType, KeyType> {
+            KeyType local_key = 0;
+            std::unordered_map<KeyType, KeyType> aux_map;
+            for (const auto &r_tup : rList) {
+                const auto *p_var_data = std::get<0>(r_tup);
+                aux_map[p_var_data->Key()] = local_key++;
+            }
+            return aux_map;
+        } (rConvergenceVariablesList))
+    {}
```

Comment:
maybe adding a check in case the user passes e.g. `VELOCITY` and `VELOCITY_X` would be good

Reply:
Just mentioning that would be nice to have the `try_emplace` feature from c++17 here :disappointed:

Reply:
Done. Now it is checked that the convergence variables  are not repeated. Also that component variables are not provided.

Reply:
I just checked again, why is checking of components not supported?
From a physical pov I would say this makes sense depending on the model